### ISSUE TITLE
LSTM slower on TITAN V vs TITAN Xp

### ISSUE THREAD

**Issue by laucheukhim** [keras-team/keras#8885]
LSTM training on Keras is significantly slower with TITAN V than that with TITAN Xp. I ran the [IMDB Bidirectional LSTM example](https://github.com/keras-team/keras/blob/master/examples/imdb_bidirectional_lstm.py) on multiple backends and found that TITAN V is at best 4% faster and at worst 147% slower than TITAN Xp.

OS: Ubuntu 16.04.3 LTS
Drivers: NVIDIA 387.34, CUDA 9.1, cuDNN 7.0.5
Python: 3.6 Anaconda

I ran the example on multiple backends, and I also ran a modified one that uses CuDNNLSTM on Tensorflow. The following results are in seconds per epoch:

<table>
  <tr>
    <th rowspan="2"></th>
    <th colspan="3">Xeon E5-1650 v4 @ 3.6GHz</th>
    <th colspan="3">Xeon E5-1680 v3 @ 4.4GHz</th>
  </tr>
  <tr>
    <th>TITAN V</th>
    <th>TITAN Xp</th>
    <th>CPU</th>
    <th>TITAN V</th>
    <th>TITAN Xp</th>
    <th>CPU</th>
  </tr>
  <tr>
    <th colspan="7">Keras 2.1.2 LSTM</th>
  </tr>
  <tr>
    <td>Tensorflow 1.5.0rc0</td>
    <td>161s</td>
    <td>140s</td>
    <td>65s</td>
    <td>268s</td>
    <td>265s</td>
    <td>45s</td>
  </tr>
  <tr>
    <td>Theano 1.0.1</td>
    <td>176s</td>
    <td>103s</td>
    <td>125s</td>
    <td>220s</td>
    <td>89s</td>
    <td>102s</td>
  </tr>
  <tr>
    <td>CNTK 2.3.1</td>
    <td>72s</td>
    <td>74s</td>
    <td>196s</td>
    <td>71s</td>
    <td>74s</td>
    <td>188s</td>
  </tr>
  <tr>
    <th colspan="7">Keras 2.1.2 CuDNNLSTM</th>
  </tr>
  <tr>
    <td>Tensorflow 1.5.0rc0</td>
    <td>23s</td>
    <td>8s</td>
    <td></td>
    <td>24s</td>
    <td>11s</td>
    <td></td>
  </tr>
  <tr>
    <th colspan="7">Keras 1.2.2 LSTM</th>
  </tr>
  <tr>
    <td>MXNet 0.12.1</td>
    <td>86s</td>
    <td>66s</td>
    <td>588s</td>
    <td>105s</td>
    <td>66s</td>
    <td>620s</td>
  </tr>
</table>

This looks like a continuation of the problem in #6640 that LSTM performance decreases with generation.

**Comment by hiandersson** [comment]
That example is not using Keras CuDNNLSTM layers (introduced in 2.0.9)? https://keras.io/layers/recurrent/#cudnnlstm
https://github.com/keras-team/keras/releases

**Comment by laucheukhim** [comment]
@hiandersson I _modified_ it, just to test, please read carefully. There are 2 separate tests I run. One with the original script, another one with `LSTM` replaced by `CuDNNLSTM`. Both tests indicate that TITAN V runs slower.

**Comment by hiandersson** [comment]
@laucheukhim read too fast (just looked at the example), my bad! Its interesting that CuDNNLSTM on TitanXp is 26 x faster than 'normal' LSTM. Is this the best overall performance you can get (comparing with Theano, CNTK)?

**Comment by laucheukhim** [comment]
@hiandersson Yes. Unfortunately, CuDNNLSTM does not support `dropout` and `recurrent_dropout`, so it cannot be used as a drop in replacement.

**Comment by hiandersson** [comment]
@laucheukhim is his a problem in Keras, tensorflow or with nvidia? is there any prognosis for a fix?

**Comment by laucheukhim** [comment]
@hiandersson I want a fix too. TITAN V is too expensive to perform badly like this.

I don't know what the cause is. I see [similar degradation in pytorch](https://discuss.pytorch.org/t/solved-titan-v-on-pytorch-0-3-0-cuda-9-0-cudnn-7-0-is-much-slower-than-1080-ti/11320/4) too. This could be due to the lack of cuDNN autotune.

TITAN V runs 15% faster in games, 75% faster in lyra2rev2 crypto mining, but 60% slower in my LSTM model. This is a big disappointment for [what is meant to be a machine learning card](https://devtalk.nvidia.com/default/topic/1027804/cuda-programming-and-performance/titan-v-tensorflow-performance/1).

**Comment by grafael** [comment]
Just for curiosity, have you tried CNTK with CuDNNLSTM ? How was that?

**Comment by laucheukhim** [comment]
@grafael CuDNNLSTM is only available on the TensorFlow backend.

**Comment by gomesraf** [comment]
Thanks @laucheukhim
